File: IntegrationTests/Parser/TypeParsing/TypeParserTests/Inheritance/InheritanceConstructor_TypeParserTest.cs
﻿// Copyright 2012 rubicon informationstechnologie gmbh
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using InjectionCop.Utilities;
using Microsoft.FxCop.Sdk;
using NUnit.Framework;

namespace InjectionCop.IntegrationTests.Parser.TypeParsing.TypeParserTests.Inheritance
{
  [TestFixture]
  public class InheritanceConstructor_TypeParserTest : TypeParserTestBase
  {
    [Test]
    public void Parse_ConstructorChainingWithBaseClassCorrectFragment_NoProblem ()
    {
      TypeNode sampleTypeNode = IntrospectionUtility.TypeNodeFactory<InheritanceSampleConstructor>();
      Method sample = IntrospectionUtility.MethodFactory (sampleTypeNode, ".ctor");
      _typeParser.Parse (sample);
      ProblemCollection result = _typeParser.Problems;

      Assert.That (TestHelper.ContainsProblemID (c_InjectionCopRuleId, result), Is.False);
    }

    [Test]
    public void Parse_ConstructorChainingWithBaseClassIncorrectFragment_ReturnsProblem ()
    {
      TypeNode stringTypeNode = IntrospectionUtility.TypeNodeFactory<string>();
      TypeNode sampleTypeNode = IntrospectionUtility.TypeNodeFactory<InheritanceSampleConstructor>();
      Method sample = IntrospectionUtility.MethodFactory (sampleTypeNode, ".ctor", stringTypeNode);
      _typeParser.Parse (sample);
      ProblemCollection result = _typeParser.Problems;

      Assert.That (TestHelper.ContainsProblemID (c_InjectionCopRuleId, result), Is.True);
    }

    [Test]
    public void Parse_ConstructorChainingWithBaseClassConstructorWithoutFragment_NoProblem ()
    {
      TypeNode stringTypeNode = IntrospectionUtility.TypeNodeFactory<string>();
      TypeNode sampleTypeNode = IntrospectionUtility.TypeNodeFactory<InheritanceSampleConstructor>();
      Method sample = IntrospectionUtility.MethodFactory (sampleTypeNode, ".ctor", stringTypeNode, stringTypeNode);
      _typeParser.Parse (sample);
      ProblemCollection result = _typeParser.Problems;

      Assert.That (TestHelper.ContainsProblemID (c_InjectionCopRuleId, result), Is.False);
    }

  }
}

File: IntegrationTests.VB/Parser/TypeParsing/TypeParserTests/Inheritance/Constructor/InheritanceConstructor_TypeParserTest.vb
' Copyright 2013 rubicon informationstechnologie gmbh
'
' Licensed under the Apache License, Version 2.0 (the "License");
' you may not use this file except in compliance with the License.
' You may obtain a copy of the License at
'
' http://www.apache.org/licenses/LICENSE-2.0
'
' Unless required by applicable law or agreed to in writing, software
' distributed under the License is distributed on an "AS IS" BASIS,
' WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
' See the License for the specific language governing permissions and
' limitations under the License.
Imports InjectionCop.Utilities
Imports Microsoft.FxCop.Sdk
Imports NUnit.Framework
Imports System

Namespace Parser.TypeParsing.TypeParserTests.Inheritance.Constructor
	<TestFixture()>
	Public Class InheritanceConstructor_TypeParserTest
		Inherits TypeParserTestBase

		<Test()>
		Public Sub Parse_ConstructorChainingWithBaseClassCorrectFragment_NoProblem()
			Dim sampleTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of InheritanceSampleConstructor)()
			Dim sample As Method = IntrospectionUtility.MethodFactory(sampleTypeNode, ".ctor", New TypeNode() {})
			Me._typeParser.Parse(sample)
			Dim result As ProblemCollection = Me._typeParser.Problems
			Assert.That(TestHelper.ContainsProblemID(Me.c_InjectionCopRuleId, result), [Is].[False])
		End Sub

		<Test()>
		Public Sub Parse_ConstructorChainingWithBaseClassIncorrectFragment_ReturnsProblem()
			Dim stringTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of String)()
			Dim sampleTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of InheritanceSampleConstructor)()
			Dim sample As Method = IntrospectionUtility.MethodFactory(sampleTypeNode, ".ctor", New TypeNode() { stringTypeNode })
			Me._typeParser.Parse(sample)
			Dim result As ProblemCollection = Me._typeParser.Problems
			Assert.That(TestHelper.ContainsProblemID(Me.c_InjectionCopRuleId, result), [Is].[True])
		End Sub

		<Test()>
		Public Sub Parse_ConstructorChainingWithBaseClassConstructorWithoutFragment_NoProblem()
			Dim stringTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of String)()
			Dim sampleTypeNode As TypeNode = IntrospectionUtility.TypeNodeFactory(Of InheritanceSampleConstructor)()
			Dim sample As Method = IntrospectionUtility.MethodFactory(sampleTypeNode, ".ctor", New TypeNode() { stringTypeNode, stringTypeNode })
			Me._typeParser.Parse(sample)
			Dim result As ProblemCollection = Me._typeParser.Problems
			Assert.That(TestHelper.ContainsProblemID(Me.c_InjectionCopRuleId, result), [Is].[False])
		End Sub
	End Class
End Namespace
